protocol/smart-contract-versioning.md

```diff
@@ -0,0 +1,86 @@
+
+- **Monorepo Releases**:  The current versioning scheme of monorepo contract releases does not follow a true semver. For example, the Fault Proofs release introduced breaking changes by changing the ABI, and we are releasing this as a minor version bump from `op-contracts/v1.3.0` to `op-contracts/v1.4.0`. This was not an intentional decision. We likely slid into this path because node releases cannot bump the major version (for go-specific reasons), and our current [contracts release process](https://www.notion.so/oplabs/Minimal-Contracts-Release-Process-ef8e91b1b13944e7a7f632a824895764?pvs=4) does not specify how to determine the version bump for an upcoming monorepo release.
+- **Individual Contracts**: The current versioning scheme of individual contracts is also not a true semver, but in a different way. We do bump major and minor versions appropriately, but we don't distinguish when a contract is in beta, release candidate, or feature status. This makes it difficult to determine the stability and compatibility of a given contract version.
+
+These have resulted in various issues, such as chains deploying in-development versions of contracts.
+
+Unlike typical software development, OP Stack chains cannot rely on "compatible semvers" to compose the chain of various contracts.
+Instead, a given monorepo contracts release must be defined by an exact set of contract versions, not a range of semver-compatible versions.
+
+# Alternatives Considered
+
+1. Keep the current versioning scheme: This alternative would maintain the status quo but would not address the issues of clarity, consistency, and alignment with semver practices.
+
+2. Use semver for contracts but not for monorepo releases: This alternative would improve the versioning of individual contracts but would not provide a clear mapping between monorepo releases and contract versions.
+
+# Proposed Solution
+
+The proposed solution is to more closely adopt a true semver format for both individual contracts and monorepo releases.
+
+We preserve the existing rules for when to bump major, minor, and patch versions in individual contracts:
+
+- `patch` releases are to be used only for changes that do NOT modify contract bytecode (such as updating comments).
+- `minor` releases are to be used for changes that modify bytecode OR changes that expand the contract ABI provided that these changes do NOT break the existing interface.
+- `major` releases are to be used for changes that break the existing contract interface OR changes that modify the security model of a contract.
+
+Versioning for individual contracts:
+
+- A contract is only `X.Y.Z` on `develop` if it's been governance approved. If it's `X.Y.Z` before that, it must be on a branch. More on this below.
+- For contracts undergoing development, a `-beta.n` identifier must be appended to the version number.
+- For contracts in a release candidate state, an `-rc.n` identifier must be appended to the version number.
```

Comment:
Does this prevent us from 'releasing the lock' on `develop` to other projects once a tag has been submitted as a governance proposal?
Or is the plan to remove the `-rc.n` suffix on a separate branch?

Reply:
The workflow would be like this:
- A PR into develop which replaces the `-beta.n` suffix with the `-rc.n` suffix for all contracts in your release. For this release, this is the last commit to develop.
  - At this point, the lock on develop is released.
- Create a branch from that commit called `proposal/op-contracts/vX.Y.Z`.
- Open a PR into that branch to remove all the `-rc.n` suffixes. Once merged, this commit is where contracts are deployed, and this commit is tagged as `op-contracts/vX.Y.Z-rc.1` for the governance proposal
- Subsequent changes only bump the `rc.n` identifier both at the contract and monorepo release level (possible exception: a change is at a higher version level than prior bumps, e.g. if this rc only bumped the minor version but now we introduce a breaking change for `rc.2` we need to bump the major version. But ideally once we're at an rc this should not be happening)
- Once the governance proposal passes, create the `op-contracts/vX.Y.Z` release from the latest `op-contracts/vX.Y.Z-rc.n` release, then merge the branch into develop

Reply:
I have incorporated this info into the primary doc in 0f9f896